src/main/java/io/confluent/connect/jdbc/sink/metadata/FieldsMetadata.java

```diff
@@ -161,19 +160,29 @@ public static FieldsMetadata extract(
           }
           keyFieldNames.addAll(configuredPkFields);
         }
+        for (String fieldName : keyFieldNames) {
+          final Schema fieldSchema = valueSchema.field(fieldName).schema();
+          allFields.put(fieldName, new SinkRecordField(fieldSchema.type(), fieldName, true, false, fieldSchema.defaultValue()));
+        }
       }
       break;
 
     }
 
+    final Set<String> nonKeyFieldNames = new LinkedHashSet<>();
     if (valueSchema != null) {
-      for (Field field : valueSchema.fields()) {
-        final boolean isKeyField = keyFieldNames.contains(field.name());
-        if (!isKeyField) {
-          nonKeyFieldNames.add(field.name());
+      for (Field field: valueSchema.fields()) {
```

Comment:
This might be easier to understand when expressed as set operations, i.e. valueSchema.fields() - keyFieldNames) and fieldsWhitelist. I haven't worked through exactly what that looks like w/ Java's set APIs, but might reduce this to a one-liner that's easier to understand.